File: SDK Examples/Examples/C#/Counter_SnapCounter/SnapCounterForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Text.RegularExpressions;
using Automation.BDaq;

namespace Counter_SnapCounter
{
    public partial class SnapCounterForm : Form
    {
		  bool m_isTimerUsing;
        int snapCount = 0;

        TimerPulseCtrl timerPulseCtrl1;

        int[] snapSrcs = {0};
		  EventId[] eventId = {0};
        ListBox.SelectedIndexCollection selectedItems;

        protected delegate void UdCntrEvent(int snapCount, UdCntrEventArgs e);
        
        public SnapCounterForm()
        {
            InitializeComponent();
            timerPulseCtrl1 = new TimerPulseCtrl();
        }

        public SnapCounterForm(int deviceNumber)
        {
            InitializeComponent();
            udCounterCtrl1.SelectedDevice = new DeviceInformation(deviceNumber);
            timerPulseCtrl1 = new TimerPulseCtrl();
        }

        protected void DataHandler(int snapCount, UdCntrEventArgs e)
        {
            if (CounterValueList.Items.Count > 40)
            {
                CounterValueList.Items.RemoveAt(0);
            }

            string snapValue;
            string srcID = "";
            string source;

            txtSnapCount.Text = snapCount.ToString();

            EventIDtoString((EventId)e.Id, ref srcID);

            snapValue = "CntrValue: ";
            for (int i = 0; i < e.Length; i++)
            {
                snapValue += e.Data[i].ToString("X");
                snapValue += " ";
            }
            CounterValueList.Items.Add(snapValue);

            source = "SrcID: ";
            source += srcID;
            CounterValueList.Items.Add(source);
        }

        private void SnapCounterForm_Load(object sender, EventArgs e)
        {
            if (!udCounterCtrl1.Initialized)
            {
                MessageBox.Show("No device be selected or device open failed!", "ContinueCompare");
                this.Close();
                return;
            }

            //set the title
            this.Text = "Snap Counter(" + udCounterCtrl1.SelectedDevice.Description + ")";

            selectedItems = listSnapSource.SelectedIndices;

            txtTimeInterval.Text = "5";
            CounterValueList.Items.Clear();
            m_isTimerUsing = false;

            ConfiguratePanel();

            btnStart.Enabled = true;
            btnStop.Enabled = false;
        }

        public void ConfiguratePanel()
        {
            //for reset value
            snapSrcs = udCounterCtrl1.Features.SnapEventSources;
				eventId = udCounterCtrl1.Features.UdSnapEventSources;
            int count = snapSrcs.Length;
            string str = "";

            for (int i = 0; i < count; i++)
            {
                EventIDtoString((EventId)snapSrcs[i], ref str);
                listSnapSource.Items.Add(str);
            }
        }

        private void HandleError(Exception err)
        {
            MessageBox.Show("Sorry ! Some errors happened, the error code is: " + err.ToString(), "Warning Information");
        }

        public void EventIDtoString(EventId id, ref string str)
        {
            switch (id)
            {
                case EventId.EvtUdIndex0:
                    str = "Index0";
                    break;
                case EventId.EvtUdIndex1:
                    str = "Index1";
                    break;
                case EventId.EvtUdIndex2:
                    str = "Index2";
                    break;
                case EventId.EvtUdIndex3:
                    str = "Index3";
                    break;
                case EventId.EvtUdIndex4:
                    str = "Index4";
                    break;
                case EventId.EvtUdIndex5:
                    str = "Index5";
                    break;
                case EventId.EvtUdIndex6:
                    str = "Index6";
                    break;
                case EventId.EvtUdIndex7:
                    str = "Index7";
                    break;
                case EventId.EvtCntTimer0:
                    str = "Timer0";
                    break;
                case EventId.EvtCntTimer1:
                    str = "Timer1";
                    break;
                case EventId.EvtCntTimer2:
                    str = "Timer2";
                    break;
                case EventId.EvtCntTimer3:
                    str = "Timer3";
                    break;
                case EventId.EvtCntTimer4:
                    str = "Timer4";
                    break;
                case EventId.EvtCntTimer5:
                    str = "Timer5";
                    break;
                case EventId.EvtCntTimer6:
                    str = "Timer6";
                    break;
                case EventId.EvtCntTimer7:
                    str = "Timer7";
                    break;
                case EventId.EvtDiintChannel000:
                    str = "DiintChannel0";
                    break;
                case EventId.EvtDiintChannel001:
                    str = "DiintChannel1";
                    break;
                case EventId.EvtDiintChannel002:
                    str = "DiintChannel2";
                    break;
                case EventId.EvtDiintChannel003:
                    str = "DiintChannel3";
                    break;
                case EventId.EvtDiintChannel004:
                    str = "DiintChannel4";
                    break;
                case EventId.EvtDiintChannel005:
                    str = "DiintChannel5";
                    break;
                case EventId.EvtDiintChannel006:
                    str = "DiintChannel6";
                    break;
                case EventId.EvtDiintChannel007:
                    str = "DiintChannel7";
                    break;
                case EventId.EvtDiCosintPort000:
                    str = "DiConsintChannel0";
                    break;
                case EventId.EvtDiCosintPort001:
                    str = "DiConsintChannel1";
                    break;
                case EventId.EvtDiCosintPort002:
                    str = "DiConsintChannel2";
                    break;
                case EventId.EvtDiCosintPort003:
                    str = "DiConsintChannel3";
                    break;
                case EventId.EvtDiPmintPort000:
                    str = "DiPmintChannel0";
                    break;
                case EventId.EvtDiPmintPort001:
                    str = "DiPmintChannel1";
                    break;
                case EventId.EvtDiPmintPort002:
                    str = "DiPmintChannel2";
                    break;
                case EventId.EvtDiPmintPort003:
                    str = "DiPmintChannel3";
                    break;

                default:
                    str = "";
                    break;
            }    
        }

        public void CheckTimerUsingParam(EventId id, ref int timerChan, ref bool isTimerUsing)
        {
            if (id >= EventId.EvtCntTimer0 && id <= EventId.EvtCntTimer7)
            {
                timerChan = id - EventId.EvtCntTimer0;
            }

            if (timerChan != -1)
            {
                isTimerUsing = true;
            }
        }

        private void timer1_Tick(object sender, EventArgs e)
        {
			  int value = 0; 
			  try
            {
                //txtCounterValue.Text = ((uint)udCounterCtrl1.Value).ToString();
					 udCounterCtrl1.Read(out value);
					 txtCounterValue.Text = value.ToString();
            }
            catch (Exception ex)
            {
                HandleError(ex);
            }
        }

        private void listSnapSource_SelectedIndexChanged(object sender, EventArgs e)
        {
            selectedItems = listSnapSource.SelectedIndices;

            int timerChan = -1;
            bool isTimerUsing = false;

            if (selectedItems.Count != 0)
            {
                foreach (int index in selectedItems)
                {
                    CheckTimerUsingParam((EventId)snapSrcs[index], ref timerChan, ref isTimerUsing);
                    if (isTimerUsing)
                    {
                        txtTimeInterval.Enabled = true;
                        break;
                    }
                    else
                    {
                        txtTimeInterval.Enabled = false;
                    }
                }
            }
            else
            {
                txtTimeInterval.Enabled = false;
            }
        }

        private void btnStop_Click(object sender, EventArgs e)
        {
            timer1.Enabled = false;

            btnStart.Enabled = true;
            btnStop.Enabled = false;
            listSnapSource.Enabled = true;

            //stop timer pulse
            if (m_isTimerUsing)
            {
                timerPulseCtrl1.Enabled = false;
                timerPulseCtrl1.Dispose();
            }

            //stop snap function
            foreach (int index in selectedItems)
            {
                //udCounterCtrl1.SnapStop(snapSrcs[index]);
					 udCounterCtrl1.SnapStop(eventId[index]);
            }

            try
            {
                //stop updown counter
                udCounterCtrl1.Enabled = false;
            }
            catch (Exception ex)
            {
                HandleError(ex);
            }
        }

        public bool isNum(string str)
        {
            bool num;
            if (str != null && Regex.IsMatch(str, @"^\d+$"))
            {
                num = true;
            }
            else
            {
                num = false;
            }
            return num;
        }

        private void btnStart_Click(object sender, EventArgs e)
        {
            CounterValueList.Items.Clear();
            txtCounterValue.Text = "0";
            txtSnapCount.Text = "0";
            snapCount = 0;

            m_isTimerUsing = false;
            int timerChan = -1;
            foreach (int index in selectedItems)
            {
                CheckTimerUsingParam((EventId)snapSrcs[index], ref timerChan, ref m_isTimerUsing);
                //udCounterCtrl1.SnapStart(snapSrcs[index]);
					 udCounterCtrl1.SnapStart(eventId[index]);
            }
            
            try
            {
                if (m_isTimerUsing)
                {
                    int deviceNumber = udCounterCtrl1.SelectedDevice.DeviceNumber;
                    DeviceInformation selectedDevice = new DeviceInformation(deviceNumber);
                    timerPulseCtrl1.SelectedDevice = selectedDevice;
                    double frequency = double.Parse(txtTimeInterval.Text);
                    timerPulseCtrl1.ChannelStart = timerChan;
                    //timerPulseCtrl1.Frequency = frequency;
						  
						  TmrChannel[] tmrChannel = timerPulseCtrl1.Channels;
						  tmrChannel[0].Frequency = frequency;
                    timerPulseCtrl1.Enabled = true;
                }
                
                //Start Snap Counter function.
                udCounterCtrl1.Enabled = true;
            }
            catch (Exception ex)
            {
                HandleError(ex);

                if (!isNum(txtTimeInterval.Text)) txtTimeInterval.Text = "";

                //stop snap function
                foreach (int index in selectedItems)
                {
                    //udCounterCtrl1.SnapStop(snapSrcs[index]);
						  udCounterCtrl1.SnapStop(eventId[index]);
                }

                return;
            }

            timer1.Enabled = true;

            btnStart.Enabled = false;
            btnStop.Enabled = true;
            listSnapSource.Enabled = false;
        }

        private void udCounterCtrl1_UdCntrEvent(object sender, UdCntrEventArgs e)
        {
            snapCount++;

            Invoke(new UdCntrEvent(DataHandler), new object[] { snapCount, e });
        }
    }
}

File: SDK Examples/Examples/VB.Net/Counter_SnapCounter/SnapCounterForm.vb
Imports System.Math
Imports Automation.BDaq

Public Class SnapCounterForm
	Inherits Form

	Private m_isTimerUsing As Boolean
	Private snapSrcs() As EventId
	Private selectedItems As ListBox.SelectedIndexCollection

	Private snapCount As Integer

	Private TimerPulseCtrl1 As TimerPulseCtrl


	Protected Delegate Sub UdCntrEvent(ByVal snapCount As Integer, ByVal e As UdCntrEventArgs)

	Public Sub New()
		InitializeComponent()
		TimerPulseCtrl1 = New TimerPulseCtrl
	End Sub

	Public Sub New(ByVal deviceNumber As Integer)
		InitializeComponent()
		UdCounterCtrl1.SelectedDevice = New DeviceInformation(deviceNumber)
		TimerPulseCtrl1 = New TimerPulseCtrl
	End Sub

	Private Sub HandleError(ByVal err As Exception)
		MessageBox.Show("Sorry ! There're some errors happened, the error code is: " & err.ToString(), "Warning Information")
	End Sub

	Private Sub EventIDtoString(ByVal id As Long, ByRef Str As String)
		Select Case id

			Case 406	  'EvtUdIndex0
				Str = "Index0"
			Case 407	  'EvtUdIndex1
				Str = "Index1"
			Case 408	  'EvtUdIndex2
				Str = "Index2"
			Case 409	  'EvtUdIndex3
				Str = "Index3"
			Case 410	  'EvtUdIndex4
				Str = "Index4"
			Case 411	  'EvtUdIndex5
				Str = "Index5"
			Case 412	  'EvtUdIndex6
				Str = "Index6"
			Case 413	  'EvtUdIndex7
				Str = "Index7"

			Case 390	  'EvtCntTimer0
				Str = "Timer0"
			Case 391	  'EvtCntTimer1
				Str = "Timer1"
			Case 392	  'EvtCntTimer2
				Str = "Timer2"
			Case 393	  'EvtCntTimer3
				Str = "Timer3"
			Case 394	  'EvtCntTimer4
				Str = "Timer4"
			Case 395	  'EvtCntTimer5
				Str = "Timer5"
			Case 396	  'EvtCntTimer6
				Str = "Timer6"
			Case 397	  'EvtCntTimer7
				Str = "Timer7"

			Case 12	  'EvtDiintChannel000
				Str = "DiintChannel0"
			Case 13	  'EvtDiintChannel001
				Str = "DiintChannel1"
			Case 14	  'EvtDiintChannel002
				Str = "DiintChannel2"
			Case 15	  'EvtDiintChannel003
				Str = "DiintChannel3"
			Case 16	  'EvtDiintChannel004
				Str = "DiintChannel4"
			Case 17	  'EvtDiintChannel005
				Str = "DiintChannel5"
			Case 18	  'EvtDiintChannel006
				Str = "DiintChannel6"
			Case 19	  'EvtDiintChannel007
				Str = "DiintChannel7"
			Case Else
				Str = ""
		End Select
	End Sub

	Private Sub CheckTimerUsingParam(ByVal id As Long, ByRef timerChan As Integer, ByRef isTimerUsing As Boolean)
		If id >= 390 And id <= 397 Then
			timerChan = id - 390
		End If

		If timerChan <> -1 Then
			isTimerUsing = True
		End If
	End Sub

	Private Sub ConfiguratePanel()
		'for reset value
		snapSrcs = UdCounterCtrl1.Features.UdSnapEventSources
		Dim srcCount As Integer
		srcCount = UBound(snapSrcs) - LBound(snapSrcs) + 1
		Dim srcSrc As String
		srcSrc = ""

		listSnapSource.Items.Clear()

		Dim i As Integer
		For i = 0 To srcCount - 1
			EventIDtoString(snapSrcs(i), srcSrc)
			listSnapSource.Items.Add(srcSrc)
		Next
	End Sub

	Private Sub btnStart_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnStart.Click
		CounterValueList.Items.Clear()

		txtCounterValue.Text = "0"
		txtSnapCount.Text = "0"
		snapCount = 0

		m_isTimerUsing = False

		Dim timerChan As Integer
		timerChan = -1

		For Each index As Integer In selectedItems
			CheckTimerUsingParam(snapSrcs(index), timerChan, m_isTimerUsing)
			UdCounterCtrl1.SnapStart(snapSrcs(index))
		Next

		If m_isTimerUsing Then
			Dim deviceNum As Integer
			deviceNum = UdCounterCtrl1.SelectedDevice.DeviceNumber
			TimerPulseCtrl1.SelectedDevice = New DeviceInformation(deviceNum)

			Dim frequency As Double
			frequency = CDbl(txtTimeInterval.Text)

			TimerPulseCtrl1.ChannelStart = timerChan
			Dim tmrChannel() As TmrChannel = TimerPulseCtrl1.Channels
			tmrChannel(0).Frequency = frequency

			Try
				TimerPulseCtrl1.Enabled = True
			Catch ex As Exception
				HandleError(ex)
			End Try
		End If

		'Start Snap Counter function.
		Try
			UdCounterCtrl1.Enabled = True
		Catch ex As Exception
			HandleError(ex)
		End Try

		Timer1.Enabled = True

		btnStart.Enabled = False
		btnStop.Enabled = True
		listSnapSource.Enabled = False
	End Sub

	Private Sub btnStop_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnStop.Click
		Timer1.Enabled = False

		btnStart.Enabled = True
		btnStop.Enabled = False
		listSnapSource.Enabled = True

		'stop timer pulse
		If m_isTimerUsing Then
			TimerPulseCtrl1.Enabled = False
			TimerPulseCtrl1.Cleanup()
		End If

		'stop snap function
		For Each index As Integer In selectedItems
			UdCounterCtrl1.SnapStop(snapSrcs(index))
		Next

		'stop updown counter
		Try
			UdCounterCtrl1.Enabled = False
		Catch ex As Exception
			HandleError(ex)
		End Try
	End Sub

	Private Sub SnapCounterForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
		'The default device of project is PCI-1784.
		If Not UdCounterCtrl1.Initialized Then
			MsgBox("Please select a device with DAQNavi wizard!", , "Warning Information")
			End
		End If

		'set title of the form
		Me.Text = "Snap Counter(" + UdCounterCtrl1.SelectedDevice.Description + ")"

		selectedItems = listSnapSource.SelectedIndices

		'initial the form
		txtTimeInterval.Text = "5"
		m_isTimerUsing = False
		snapCount = 0
		CounterValueList.Items.Clear()

		ConfiguratePanel()

		btnStart.Enabled = True
		btnStop.Enabled = False
	End Sub

	Private Sub listSnapSource_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles listSnapSource.SelectedIndexChanged
		selectedItems = listSnapSource.SelectedIndices

		Dim timerChan As Integer
		timerChan = -1
		Dim isTimerUsing As Boolean
		isTimerUsing = False

		If selectedItems.Count <> 0 Then
			For Each index As Integer In selectedItems
				CheckTimerUsingParam(snapSrcs(index), timerChan, isTimerUsing)

				If isTimerUsing Then
					txtTimeInterval.Enabled = True
					Exit For
				Else
					txtTimeInterval.Enabled = False
				End If
			Next
		Else
			txtTimeInterval.Enabled = False
		End If
	End Sub

	Private Sub Timer1_Tick(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Timer1.Tick
		Try
			UdCounterCtrl1.Enabled = True
		Catch ex As Exception
			Timer1.Enabled = False
			HandleError(ex)
		End Try

		Dim cntrValue As Integer
		UdCounterCtrl1.Read(cntrValue)

		txtCounterValue.Text = "0x" + Hex(cntrValue)
	End Sub

	Private Sub UdCounterCtrl1_UdCntrEvent(ByVal sender As System.Object, ByVal e As Automation.BDaq.UdCntrEventArgs) Handles UdCounterCtrl1.UdCntrEvent
		snapCount = snapCount + 1

		Try
			Me.Invoke(New UdCntrEvent(AddressOf DataHandler), New Object() {snapCount, e})
		Catch ex As System.Exception
		End Try
	End Sub

	Private Sub DataHandler(ByVal snapCount As Integer, ByVal e As UdCntrEventArgs)
		If CounterValueList.Items.Count > 40 Then
			CounterValueList.Items.RemoveAt(0)
		End If

		Dim snapValue As String
		Dim srcID As String
		srcID = ""
		Dim source As String

		txtSnapCount.Text = Str(snapCount)

		EventIDtoString(e.Id, srcID)

		snapValue = "CntrValue: "
		Dim i As Integer
		For i = 0 To e.Length
			snapValue = snapValue & Hex(e.Data(i))
		Next
		CounterValueList.Items.Add(snapValue)

		source = "SrcID: "
		source = source & srcID
		CounterValueList.Items.Add(source)
	End Sub
End Class
